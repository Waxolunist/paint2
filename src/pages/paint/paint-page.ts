import {
  customElement,
  LitElement,
  html,
  css,
  property,
  query,
} from 'lit-element';
import {arrowBack} from './icons';
import '../../components/icon-button/paint-icon-button';
import '../../components/color-toolbar/paint-color-toolbar';
import '../../components/paint-area/paint-area';
import store from '../../store';
import {navigate} from 'lit-redux-router';

@customElement('paint-paint-page')
export class PaintPage extends LitElement {
  @query('[name="paint-area-container"]')
  private area?: HTMLElement;

  @property({type: Number, reflect: false})
  width = 0;

  @property({type: Number, reflect: false})
  height = 0;

  @property({type: String, reflect: false})
  colorCode = '#000';

  firstUpdated() {
    this.calcAspectRatio();
  }

  isLandscape({height, width}: {height?: number; width?: number} = {}) {
    if (width && height) {
      return width > height;
    }
    return window.screen.orientation.type.startsWith('landscape');
  }

  calcAspectRatio() {
    const {height, width} = this.area!.getBoundingClientRect();
    if (this.isLandscape({height, width})) {
      console.log('landscape');
      const scale = Math.min(width / 297, height / 210);
      this.height = (210 * scale) | 0;
      this.width = (297 * scale) | 0;
    } else {
      console.log('portrait');
      const scale = Math.min(width / 210, height / 297);
      this.height = (297 * scale) | 0;
      this.width = (210 * scale) | 0;
    }
  }

  static get styles() {
    // language=CSS
    return css`
      :host {
        display: block;
        display: flex;
        height: 100vh;
        width: 100vw;
        overflow: hidden;
      }

      .paint-page {
        display: flex;
        width: 100%;
      }

      .paint-toolbar {
        width: 45px;
        min-width: 45px;
        background-color: #eee;
        display: flex;
        align-items: flex-start;
        padding-top: 5px;
        flex-direction: column;
        box-sizing: border-box;
      }

      .back-button {
        height: var(--icon-size);
        width: var(--icon-size);
        margin-bottom: 5px;
        margin-left: 7px;
      }

      .paint-wrapper {
        display: flex;
        align-items: stretch;
        flex-grow: 1;
        padding: 30px;
      }

      .paint-area-container {
        flex: 1;
        display: flex;
        justify-content: center;
        align-items: center;
      }
    `;
  }
  //
  render() {
    return html`
      <div class="paint-page">
        <div class="paint-toolbar">
          <paint-icon-button
            class="back-button"
            @icon-clicked="${this.navigateBack}"
          >
            ${arrowBack}
          </paint-icon-button>
          <paint-color-toolbar
            @color-changed="${this.colorChanged}"
          ></paint-color-toolbar>
        </div>
        <div class="paint-wrapper">
          <div name="paint-area-container" class="paint-area-container">
            <paint-area
              width="${this.width}"
              height="${this.height}"
              colorCode="${this.colorCode}"
            ></paint-area>
          </div>
        </div>
      </div>
    `;
  }

  private navigateBack() {
    store.dispatch(navigate('/'));
  }

  private colorChanged(e: CustomEvent) {
    this.colorCode = e.detail.code;
  }
}
