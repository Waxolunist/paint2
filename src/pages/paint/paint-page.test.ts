import {fixture, html} from '@open-wc/testing';
import {cleanHTML} from '../../test/htmlutils';
import {PaintPage} from './paint-page';

//jest.mock('web-worker:./paint.worker.ts')

describe('paint-page', () => {
  it('is defined', () => {
    const el = document.createElement('paint-paint-page');
    expect(el).toBeInstanceOf(PaintPage);
  });

  it('renders correctly', async () => {
    const element = await fixture(html`<paint-paint-page></paint-paint-page>`);
    expect(cleanHTML(element)).toMatchSnapshot();
  });
});
