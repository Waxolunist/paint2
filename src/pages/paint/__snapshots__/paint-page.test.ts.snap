// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`paint-page renders correctly 1`] = `
"
      <div class=\\"paint-page\\">
        <div class=\\"paint-toolbar\\">
          <paint-icon-button class=\\"back-button\\">
            <svg xmlns=\\"http://www.w3.org/2000/svg\\" viewBox=\\"0 0 448 512\\">
  <path d=\\"M257.5 445.1l-22.2 22.2c-9.4 9.4-24.6 9.4-33.9 0L7 273c-9.4-9.4-9.4-24.6 0-33.9L201.4 44.7c9.4-9.4 24.6-9.4 33.9 0l22.2 22.2c9.5 9.5 9.3 25-.4 34.3L136.6 216H424c13.3 0 24 10.7 24 24v32c0 13.3-10.7 24-24 24H136.6l120.5 114.8c9.8 9.3 10 24.8.4 34.3z\\"></path>
</svg>
          </paint-icon-button>
          <paint-color-toolbar activecolor=\\"#000000\\"></paint-color-toolbar>
        </div>
        <div class=\\"paint-wrapper\\">
          <div id=\\"paint-area-container\\" class=\\"paint-area-container\\">
            
          </div>
        </div>
      </div>
    <style>
      :host {
        display: block;
        display: flex;
        height: 100vh;
        width: 100vw;
        overflow: hidden;
      }

      .paint-page {
        display: flex;
        width: 100%;
      }

      .paint-toolbar {
        width: 45px;
        min-width: 45px;
        background-color: #eee;
        display: flex;
        align-items: flex-start;
        padding-top: 5px;
        flex-direction: column;
        box-sizing: border-box;
        position: fixed;
        top: 0;
        bottom: 0;
      }

      .back-button {
        height: var(--icon-size);
        width: var(--icon-size);
        margin-bottom: 5px;
        margin-left: 7px;
      }

      .paint-wrapper {
        display: flex;
        align-items: stretch;
        flex-grow: 1;
        padding: 30px;
        margin-left: 45px;
      }

      .paint-area-container {
        flex: 1;
        display: flex;
        justify-content: center;
        align-items: center;
      }
    </style>"
`;
